src/stan/services/optimize/bfgs.hpp

```diff
@@ -0,0 +1,200 @@
+#ifndef STAN_SERVICES_OPTIMIZE_BFGS_HPP
+#define STAN_SERVICES_OPTIMIZE_BFGS_HPP
+
+#include <stan/io/var_context.hpp>
+#include <stan/io/chained_var_context.hpp>
+#include <stan/callbacks/interrupt.hpp>
+#include <stan/io/random_var_context.hpp>
+#include <stan/callbacks/writer.hpp>
+#include <stan/services/error_codes.hpp>
+#include <stan/optimization/bfgs.hpp>
+#include <stan/services/util/initialize.hpp>
+#include <stan/services/util/rng.hpp>
+#include <fstream>
+#include <iostream>
+#include <iomanip>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace services {
+    namespace optimize {
+
+      /**
+       * Runs the BFGS algorithm for a model.
+       *
+       * @tparam Model A model implementation
+       *
+       * @param model Input model to test (with data already instantiated)
+       * @param init var context for initialization
+       * @param random_seed random seed for the pseudo random number generator
+       * @param chain chain id to advance the pseudo random number generator
+       * @param init_radius radius to initialize
+       * @param[in] init_alpha
+       * @param[in] tol_obj
+       * @param[in] tol_rel_obj
+       * @param[in] tol_grad
+       * @param[in] tol_rel_grad
+       * @param[in] tol_param
+       * @param[in] num_iterations maximum number of iterations
+       * @param[in] save_iterations indicates whether all the interations should
```

Comment:
Should say something like "if true, values at each iteration are saved in addition to the ifnal result, otherwise only the final result is saved." Currently unclear what happens when false (does it save anything?)